### PR TITLE
Implement canvas

### PR DESCRIPTION
**Pull request by Gujansko**
Repo had some pushes while I was working in the files. If any deletions or changes are not actually from the current state of the main repository please mark them so I could change it accordingly. I followed what ide showed me was the most recent change.

### CODE REVIEW

**Comment by kowallus** [comment]
> Repo had some pushes while I was working in the files. If any deletions or changes are not actually from the current state of the main repository please mark them so I could change it accordingly. I followed what ide showed me was the most recent change.

In general, You should take responsibility for merging changes from the main repo. It is vital to verify that the newest changes are incorporated and that all features work correctly together (although, in the current case the lack of automated tests makes it very hard).

**Comment by Gujansko** [comment]
I took care of unnecessary deletions only one strange thing is that sync to main added setUpMouseHandler function to my branch even though I didn't touch mouse hanlders tab. Maybe it got added and then removed by somebody else by accident.
image

File: src/main/java/edu/kis/powp/jobs2d/command/DriverCommand.java [hunk 1]
```diff
@@ -1,17 +1,32 @@
...
 
     /**
      * Execute command on driver.
      *
      * @param driver driver.
      */
-    public void execute(Job2dDriver driver);
-    public void accept(CommandVisitor commandVisitor);
+    public void execute(Job2dDriver driver) {
+        Canvas currentCanvas = CanvasFeature.getCanvasManager().getCurrentCanvas();
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/DriverCommand.java]: Changing DriverCommands is not an option. Especially, in the case of this task, we should focus on the work of drivers (which represent API to a device), not on higher-lever, and abstract representations of commands.

Moreover, introducing canvas abstraction to such interface results in breaking SRP and OCP rules.

File: src/main/java/edu/kis/powp/jobs2d/command/canvas/Canvas.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package edu.kis.powp.jobs2d.command.canvas;
+
+
+import edu.kis.powp.jobs2d.command.CompoundCommand;
+
+public interface Canvas {
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/canvas/Canvas.java:6]: I like this interface!

File: src/main/java/edu/kis/powp/jobs2d/command/canvas/CanvasA4.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+                .addOperateTo(-WIDTH / 2, HEIGHT / 2)
+                .addOperateTo(WIDTH / 2, HEIGHT / 2)
+                .addOperateTo(WIDTH / 2, -HEIGHT / 2)
+                .addOperateTo(-WIDTH / 2, -HEIGHT / 2);
+        return builder.build();
+    }
+
+    @Override
+    public boolean isInsideCanvas(int x, int y) {
+        return x >= -WIDTH / 2 && x <= WIDTH / 2 && y >= -HEIGHT / 2 && y <= HEIGHT / 2;
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/canvas/CanvasA4.java]: Code duplication in CanvasA3 and CanvasA4 classes!

File: src/main/java/edu/kis/powp/jobs2d/features/CanvasFeature.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    public static void setupCanvas(Application application) {
+        app = application;
+        app.addComponentMenu(CanvasFeature.class, "Canvas");
+        app.addComponentMenuElement(CanvasFeature.class, "Draw Canvas",e -> {
+            if (canvasManager.getCurrentCanvas() == null) {
+                logger.warning("Canvas is not set");
+                return;
+            }
+            canvasManager.getCurrentCanvas().getCanvasCommand().execute(DriverFeature.getDriverManager().getCurrentDriver());
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/features/CanvasFeature.java]: A very long one-liner. I suggest writing this part using at least 2 lines.

**Review by kowallus — CHANGES_REQUESTED** [review]
I like the part with Canvas abstraction, manager, and feature.

Revert changes to commands hierarchy, because they have bad design, and do not reflect the task "Driver for avoiding exceeding device canvas.". You should rather implement new Job2dDriver (or driver extension using e.g., decorator approach).

File: src/main/java/edu/kis/powp/jobs2d/command/ICompoundCommand.java [hunk 5]
```diff
@@ -6,7 +6,5 @@
  * Interface extending Job2dDriverCommand to execute more than one command.
  */
 public interface ICompoundCommand extends DriverCommand {
-
-    public Iterator<DriverCommand> iterator();
-
+    Iterator<DriverCommand> iterator();
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/ICompoundCommand.java]: please revert those changes

**Review by kowallus — CHANGES_REQUESTED** [review]
please fix unnecessary minor changes in unrelated code

File: src/main/java/edu/kis/powp/jobs2d/drivers/CanvasAwareDriver.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+import edu.kis.legacy.drawer.panel.DrawPanelController;
+import edu.kis.legacy.drawer.shape.ILine;
+import edu.kis.powp.jobs2d.Job2dDriver;
+import edu.kis.powp.jobs2d.command.canvas.Canvas;
+import edu.kis.powp.jobs2d.features.CanvasFeature;
+
+import java.util.logging.Logger;
+
+public class CanvasAwareDriver implements Job2dDriver {
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/drivers/CanvasAwareDriver.java:11]: Decorator approach supports OCP and would be much more flexible (and succinct).

**Review by kowallus — APPROVED** [review]
Thank you

**Pull request merged**